change ZAP jobs to be per project

afeld opened the pull request:
Corresponds to https://trello.com/c/8Oj9KDVh/108-increase-throughput-of-scanning-projects-with-zap.

This pull request changes the ZAP pipeline from having one `scheduled` and one `ondemand` job, each of which loops through the projects and runs the (slow) ZAP scans in serial, to having `scheduled` and `ondemand` jobs per project, each under a different group. Maybe easier to understand visually:

![screen shot 2016-03-30 at 12 28 59 am](https://cloud.githubusercontent.com/assets/86842/14132010/e73f8dc8-f60e-11e5-8fb5-016e5790d275.png)
## Background

As I was thinking about how to implement [per-project Slack notifications](https://trello.com/c/AGdbPw7P/74-as-a-user-i-want-to-be-notified-when-a-zap-scan-is-complete), I was annoyed that we couldn't use the off-the-shelf [Slack resource](https://github.com/Nopik/slack-notification-resource), because it requires the channel name to be included in the pipeline. This wouldn't work for us as part of the existing ZAP pipeline, since this feature requires posting to a different channel for each project.

After [soliciting some feedback](https://github.com/concourse/concourse/issues/292) about how to handle these "multi-object pipelines", I realized that our project list a) isn't terribly huge and b) doesn't change that often, so if there were a way to have each project listed as a separate job/pipeline, we could go from having multi-object jobs (i.e. looping through the projects at run time) to single-object ones.
## Pros
- Since each job now corresponds to a single project, ZAP scans will be run in parallel.
- The jobs/tasks are now simpler, since they only deal with one project, rather than multiple.
- We can use more of the [pre-built resources](http://concourse.ci/resource-types.html), rather than needing to build our own.
- If we want, we can use task statuses more accurately (red means there was a vulnerability found), giving us a poor man's version of the Compliance Viewer.
  - Clicking on a group would show as red if there are vulnerabilities, green if there were.
  - The build history would then show the progression of a single project (e.g. red, red, green, red).
## Cons
- The ZAP pipeline now requires a manual build before uploading it.
- The Pipeline YAML is now more difficult to follow.
- The group list in the UI is going to get long very quickly.
  - Luckily, it wraps to the page, so it won't be a problem until we get in the dozens of projects.
    
    ![screen shot 2016-03-30 at 12 53 09 am](https://cloud.githubusercontent.com/assets/86842/14143492/9201ce70-f659-11e5-977b-21add7bc3530.png)
## TODOs

Will take care of these if the new structure seems agreeable to everyone.
- [ ] Test the refactored pipeline/tasks thoroughly
- [x] Clean up / split existing tasks to work with new format
- [x] Use data from the Team API to supplement the `targets.json` (like it was before)
- [x] Ensure that S3 results are being uploaded properly
- [x] Ensure that Slack notification(s) still look as expected
- [ ] **Later:** Only fetch the relevant `last-results` file, rather than all of them (should allow us to use the provided [S3 Resource](https://github.com/concourse/s3-resource))
## Conclusion

For a handful of reasons (mostly the group list outgrowing the UI), this is probably only a medium-term solution. That being said, I think the simplification is worth it for now, and we can figure something else out later.

@dlapiduz @LinuxBozo Interested to hear your feedback! Let me know if you need any more context.

afeld committed:
templatize ZAP pipeline

Instead of having one job in a pipeline that scans all of the projects, this
creates a job per project, so that they can

* be run in parallel
* have simpler pipelines
* be able to use provided Resources rather than needing to customize
* each one to use handle multiple "objects" (projects, Slack channels, etc.)

More background information:
https://github.com/concourse/concourse/issues/292

afeld commented:
I should also note: I also considered generating one pipeline per project rather than one per job. This would have avoided the group list crowding problem (the list of pipelines scrolls) and would allow us to pause scanning of individual projects (by pausing the pipeline), but I assume the cloud.gov folks would be a bit annoyed at us devouring their list of pipelines. Short of having a dedicated Concourse for Compliance Toolkit (which I am not actually suggesting), I think the option in this pull request is a better choice.

DavidEBest commented:
Interesting, @afeld. I was going to suggest a different approach: 
Have the `filter-projects` task divvy up the projects into `x` json files.
Create `x` copies of the `run-zap` job that can run in parallel, each to consume a single project.json file.

It'd give us the ability to scale out the number of parallelized buckets (increasing `x`) without going to 1:1 project parity, which (like you say) is more of a mid-term solution. It'd also allow us to 'fan-out' and then 'fan-in' to handle the summarization & s3 uploading, instead of requiring each project-level job handle that individually. It certainly isn't as obvious as a 1:1 solution, and it'd still require us to use custom resources.

Thoughts?

LinuxBozo commented:
One way to solve the "Pipeline YAML is now more difficult to follow" problem is to generate it from several yamls.. Maybe something along the lines of:

```
...
|- c2-jobs.yml
|- c2-resources.yml
...
```

`c2-jobs.yml` becomes just:

```
- name: zap-ondemand-c2
  serial: false
  plan:
    ...
- name: zap-scheduled-c2
  serial: false
  plan: 
    ...
```

Same treatment gets applied for the resources bits for each project.
You can write a script to pull it all together into a `full-pipeline.yml` or the like. 

Another alternative to the grouping by project, is grouping by the two job types, scheduled and on-demand, keeping the job per project approach. Given Concourse ability to zoom/scroll, the larger groups of jobs shouldn't be much of an issue I don't think.

DavidEBest commented:
Another benefit of the bucket approach is that we won't need to rebuild the pipeline everytime we add or remove a project, although that will be less of an issue once we get a CD pipeline for the pipeline. :)

That said, I'm not against the idea of going 1:1 project:job for the time being. It is the solution that most cleanly fits the Concourse model, even if it seems non-optimal.

ctro commented:
The thing I like most about @afeld's proposal is that it seems like simplest, straightest path with the least "cognitive load" (<-- just learned that term).  I will almost always vote for simple and straight :)

I also like the visibility that this provides into each project's runs.

Somehow this feels more in the spirit of Concourse than what we're currently doing.

Re: the Con of requiring one more "build" step, I was thinking we should put those `fly set-pipeline` commands that are currently in comments (and now the build step too) into `.sh` files that we can just run.

It also seems like the problem of too many group names across the top could be most easily solved by a PR to Concourse that put those things in some other sort of list like a dropdown.

After my short foray into custom Resources,  I'd say that being able to use pre-built resources is a HUGE gain.

afeld commented:
> Have the filter-projects task divvy up the projects into `x` json files...

Yeah, that seems like it would help with the UI crowding problem, but not with the custom resource problem, and would be even more complex. Having more Concourse groups doesn't cost us anything, so I don't think that would be worth it.

> 'fan-in' to handle the...s3 uploading

Isn't that per-project anyway?

> One way to solve the "Pipeline YAML is now more difficult to follow" problem is to generate it from several yamls

Great idea, thanks! That could also help with the duplication between the `zap-scheduled` and `zap-ondemand` jobs in the template.

> Another alternative to the grouping by project, is grouping by the two job types

Interesting. I kinda prefer it this way since it's easy to see all the information about a single project in one place...anyone else have strong feelings?

> we should put those `fly set-pipeline` commands that are currently in comments (and now the build step too) into `.sh` files that we can just run.

:+1:

> It also seems like the problem of too many group names across the top could be most easily solved by a PR to Concourse that put those things in some other sort of list like a dropdown.

Ha, I wrote this exact same thing in my response draft before I got down to reading yours. GREAT MINDS :bowtie:

---

@davidebest Let me know if you're ok with me continuing in this direction.

DavidEBest commented:
@afeld Re: the 'fan-in' - it works as is with the custom resources, and I could imagine it being more efficient to not have every job requesting the last scan, running the comparison task, and then uploading to S3, but I'm not too concerned with it.

Totally okay with this approach. Go forth and make it awesome.

afeld committed:
de-duplicate the build plans for scheduled and ondemand ZAP jobs

afeld committed:
add disclaimers to the top of the ZAP template and result pipeline files

afeld committed:
put shared library/config in central locations

afeld committed:
add Rakefile to run all minispec tests

afeld committed:
create a filter-project-data task

afeld committed:
remove creation/uploading of summary.json

Not currently used by any of our systems, and will be more complicated
to do with the parallel jobs for ZAP. More discussion here:

https://18f.slack.com/archives/compliance-toolkit/p1459540350000957

afeld committed:
add a script to update the ZAP pipeline

afeld committed:
modify summarize-zap-results to only deal with a single project

afeld commented:
Still doing some prodding at this, but I think it's ready for a detailed review! You can see the new Slack notification format [here](https://18f.slack.com/archives/ct-bot-attack/p1459754551000009). I'm keeping the TODOs up at the top up-to-date. Despite how large this pull request is, I did make a concerted effort to change as little as possible to make the new format work...capturing some refactoring ideas as Later TODOs up top.

DavidEBest commented on tasks/filter-project-data/README.md:
```diff
@@ -0,0 +1,52 @@
+  ],
+  "otherfield": "something"
+  // ...
+}
```
This example just looks like a merge, not an overwrite. To emphasize the overwrite, care to add a `links` section to `project.json`?

afeld replied:
Updated!

afeld committed:
refactor the ZAP pipeline builder

ctro commented on tasks/fetch-last-zap-results.yml:
```diff
@@ -1,16 +1,16 @@
   path: "cp"
   args:
-  - -rv
+  - -Rv
```
good catch.

ctro commented:
Wow, awesome.  So much simpler.
- Thank you for Rake (in 7 lines :))
- The ERB implementation makes sense and is simple too.  De-duping so much of the pipeline is a nice benefit!
- Such Simple for `zap/task.sh`!
- HOLY HINDSIGHT wrt. the summaries!!!  💯

ctro commented on pipelines/zap/pipeline.yml:
```diff
@@ -1,7 +1,12 @@
-#
-#   fly set-pipeline -t lite -n -c pipelines/zap.yml -p zap --load-vars-from config/local.yml -v script-branch=`git rev-parse --abbrev-ref HEAD`
-#
+<%# note that this pipeline file is actually an ERB template, and thus shouldn't be used with `fly` directly -%>
```
ha, 🎩 ✨  this is my favorite.  Now you see me...
& making it obvious (`<%`) that ERB lies ahead.

afeld committed:
make overrides in targets.json more clear

DavidEBest commented on README.md:
```diff
@@ -33,10 +33,11 @@ Running pipelines on ci.cloud.gov is identical to running pipelines locally, wit
 cp config/prod.example.yml config/prod.yml
 # modify config/prod.yml
-fly set-pipeline -t cloud -n -c pipelines/zap.yml -p zap --load-vars-from config/prod.yml
+# build the ZAP pipeline
```
Any reason not to include the command or point to the directions in the zap readme here?

DavidEBest commented on pipelines/zap/pipeline.yml:
```diff
@@ -43,69 +48,35 @@ resources:
+    file: scripts/tasks/filter-project-data/task.yml
   - task: fetch-last-zap-results
     file: scripts/tasks/fetch-last-zap-results.yml
   - task: scan-zap
```
It'd be nice to have some sort of notification that the pipeline failed, but we certainly don't want to be notified for _every single job_.

afeld replied:
Mind if we tackle that separately?

DavidEBest replied:
Works for me.

ctro commented:
I just upgraded Concourse and ran this pr locally.
Works great, I had 0 issues. 

Upgrading concourse was as easy as `vagrant box update`

DavidEBest commented:
Some thoughts:
- I like that targets.json was moved out of the specific task folder and into the general config folder.
- I like the `update` script. Another good idea to make the project more user friendly.
- This doesn't let you run it against projects that don't have a team-api entry. Not really our problem, but there are a few jobs (`confidential-survey` and `micropurchase`) that do not have team api entries, despite having about.yml files in the repo.

afeld commented:
> there are a few jobs (confidential-survey and micropurchase) that do not have team api entries, despite having about.yml files in the repo

Yeeeeah, I spent a _lot_ of time yesterday trying to figure out a) if there's a good way to handle Resource 404s, and b) [why those entries aren't appearing in the Team API](https://github.com/18F/team-api-server/issues/12). Kinda failed on both. Would you mind if I create a card for it and tackle it separately? Would prefer to get this pull request in sooner than later to be able to make smaller increments on it.

DavidEBest commented:
@afeld Seems reasonable to split that out, and I'd also like to get this in sooner rather than later. :+1:

afeld committed:
Merge remote-tracking branch 'origin/master' into zap-single-B

afeld committed:
simplify cloud.gov deployment instructions

afeld committed:
fixes for reek

afeld committed:
Merge remote-tracking branch 'origin/master' into zap-single-B

afeld committed:
add simple tests for the PipelineBuilder

afeld commented:
This is ready for a final pass!

ctro commented on pipelines/zap/update:
```diff
@@ -0,0 +1,14 @@
+fly set-pipeline -t lite -n -c tmp/zap-pipeline.yml -p zap --load-vars-from config/local.yml -v script-branch="$BRANCH"
+fly -t lite unpause-pipeline --pipeline zap
+
+open http://192.168.100.4:8080/pipelines/zap
```
Thank you for that.

ctro commented:
This still looks great to me.  :+1:  I say merge!
